Source/Core/DolphinQt2/Config/GeckoCodeWidget.cpp

```diff
@@ -0,0 +1,179 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "DolphinQt2/Config/GeckoCodeWidget.h"
+
+#include <QLabel>
+#include <QListWidget>
+#include <QMessageBox>
+#include <QPushButton>
+#include <QTextEdit>
+#include <QVBoxLayout>
+
+#include "Common/FileUtil.h"
+#include "Core/ConfigManager.h"
+#include "Core/GeckoCodeConfig.h"
+#include "DolphinQt2/Config/CheatWarningWidget.h"
+#include "DolphinQt2/GameList/GameFile.h"
+
+GeckoCodeWidget::GeckoCodeWidget(const GameFile& game)
+    : m_game(game), m_game_id(game.GetGameID().toStdString()), m_game_revision(game.GetRevision())
+{
+  CreateWidgets();
+  ConnectWidgets();
+
+  auto game_ini_default = SConfig::GetInstance().LoadDefaultGameIni(m_game_id, m_game_revision);
+  m_game_ini_local = SConfig::GetInstance().LoadLocalGameIni(m_game_id, m_game_revision);
+
+  m_gecko_codes = Gecko::LoadCodes(game_ini_default, m_game_ini_local);
+
+  UpdateList();
+}
+
+void GeckoCodeWidget::CreateWidgets()
+{
+  auto* layout = new QVBoxLayout;
+  m_warning = new CheatWarningWidget(m_game_id);
+  m_code_list = new QListWidget;
+  m_info_label = new QLabel;
+  m_code_description = new QTextEdit;
+  m_code_view = new QListWidget;
+  m_download_codes = new QPushButton(tr("Download Codes (WiiRD Database)"));
+
+  m_code_description->setReadOnly(true);
+
+  m_download_codes->setEnabled(!m_game_id.empty());
+
+  layout->addWidget(m_warning);
+  layout->addWidget(m_code_list, 3);
+  layout->addWidget(m_info_label);
+  layout->addWidget(m_code_description);
+  layout->addWidget(m_code_view);
+  layout->addWidget(m_download_codes, Qt::AlignRight);
```

Comment:
If you're using the layout's alignment property, you also need to set the widget's horizontal size policy (probably to `Fixed` or something like that).